code/streaming/Hacl.Streaming.Blake2.Common.fst

```diff
@@ -139,77 +147,45 @@ let stateful_blake2 (a : alg) (m : m_spec) : I.stateful unit =
 inline_for_extraction noextract
 let key_size (a : alg) = kk:nat{kk <= Spec.max_key a}
 
-inline_for_extraction noextract
-let key_size_t (a : alg) =
-  key_size:U32.t{U32.v key_size <= Spec.max_key a}
-
 /// Defining stateful keys
 inline_for_extraction noextract
-let stateful_key_t (a : alg) (key_size : key_size a) : Type =
-  if key_size = 0 then unit else b:B.buffer uint8 { B.length b == key_size}
-
-inline_for_extraction noextract
-let buffer_to_stateful_key_t (a : alg) (kk : key_size a{kk > 0})
-                             (k : B.buffer uint8 { B.length k == kk }) :
-  Tot (stateful_key_t a kk) =
-  k
+let stateful_key_t (a : alg) (kk: key_size_t a): Type =
+  singleton kk & (b:B.buffer uint8 { B.len b == kk})
 
+/// Keeps the key and its length at run-time, for extraction, but makes sure the
+/// parameter is ghost to let the functor decide, via the index, which APIs need
+/// a run-time key and which don't.
 inline_for_extraction noextract
-let unit_to_stateful_key_t (a : alg) :
-  Tot (stateful_key_t a 0) =
-  ()
-
-/// The ``has_key`` parameter is meta
-/// TODO: this definition could be moved to Hacl.Streaming.Interface, it could
-/// be pretty useful in other situations as it generalizes ``stateful_buffer`` in
-/// the case the length is zero. Note that rather than being unit, the type could
-/// be buffer too (and we would use null whenever needed).
-inline_for_extraction noextract
-let stateful_key (a : alg) (kk : key_size a) :
-  I.stateful unit =
+let stateful_key (a : alg):
+  I.stateful (key_size_t a) =
   I.Stateful
-    (fun _ -> stateful_key_t a kk)
+    (fun kk -> stateful_key_t a kk)
     (* footprint *)
-    (fun #_ h s -> if kk = 0 then B.loc_none else B.loc_addr_of_buffer (s <: B.buffer uint8))
+    (fun #kk h s -> if kk = 0ul then B.loc_none else B.loc_addr_of_buffer ((snd s) <: B.buffer uint8))
```

Comment:
Since we now always have a buffer in the state, do we need the disjunction here ?

Reply:
Yes because otherwise some lemmas cannot be proven for B.null (loc_in_footprint and stuff like that).

It's a massive pain and it's even worse in my upcoming PR but I don't have any way around that

Reply:
(Maybe it exists -- I just couldn't get it to work)